### PR TITLE
[IMP] required multi session for POS

### PR DESCRIPTION
**Pull request by GabbasovDinar**

### CODE REVIEW

File: pos_multi_session/models/pos_multi_session_models.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 class PosConfig(models.Model):
     _inherit = 'pos.config'
 
-    multi_session_id = fields.Many2one('pos.multi_session', 'Multi-session', help='Set the same value for POSes where orders should be synced. Keep empty if this POS should not use syncing. Before updating it you need to close active session')
+    multi_session_id = fields.Many2one('pos.multi_session', 'Multi-session', required=True, help='Set the same value for POSes where orders should be synced. Keep empty if this POS should not use syncing. Before updating it you need to close active session')
```
Comment (yelizariev) [inline pos_multi_session/models/pos_multi_session_models.py]: Fix this note:

     Keep empty if this POS should not use syncing

**Review by yelizariev — CHANGES_REQUESTED** [review]

**Review by yelizariev — CHANGES_REQUESTED** [review]
Update Version to 4.1.0

File: pos_multi_session/models/pos_multi_session_models.py [hunk 2]
```diff
@@ -16,7 +16,8 @@
 class PosConfig(models.Model):
     _inherit = 'pos.config'
 
-    multi_session_id = fields.Many2one('pos.multi_session', 'Multi-session', help='Set the same value for POSes where orders should be synced. Keep empty if this POS should not use syncing. Before updating it you need to close active session')
+    multi_session_id = fields.Many2one('pos.multi_session', 'Multi-session', required=True, help='Set the same value for POSes where orders should be synced. Deselect checkbox of "Active" if this POS should not use syncing. Before updating it you need to close active session')
```
Comment (yelizariev) [inline pos_multi_session/models/pos_multi_session_models.py]: Put required to the form instead

```
2018-05-22 11:37:27,076 25268 WARNING 16780-679-04dd3a-base odoo.models: WARNING: unable to set column multi_session_id of table pos_config not null !
Try to re-run: openerp-server --update=module
If it doesn't work, update records and execute manually:
ALTER TABLE pos_config ALTER COLUMN multi_session_id SET NOT NULL
Traceback (most recent call last):
  File "/mnt/odoo-extra/runbot/static/build/16780-679-04dd3a/odoo/models.py", line 2537, in _auto_init
    cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" SET NOT NULL' % (self._table, name))
  File "/mnt/odoo-extra/runbot/static/build/16780-679-04dd3a/odoo/sql_db.py", line 154, in wrapper
    return f(self, *args, **kwargs)
  File "/mnt/odoo-extra/runbot/static/build/16780-679-04dd3a/odoo/sql_db.py", line 231, in execute
    res = self._obj.execute(query, params)
IntegrityError: column "multi_session_id" contains null values

```

**Review by yelizariev — CHANGES_REQUESTED** [review]
Also, there should be Default Session for normal (non-demo installation)

**Comment by yelizariev** [comment]
```
************* Module pos_multi_session
pos_multi_session/multi_session_view.xml:7: [W7939(deprecated-data-xml-node), ]  Use <odoo> instead of <odoo><data> or use <odoo noupdate="1">instead of <odoo><data noupdate="1">
pos_multi_session/demo/demo.xml:8: [W7939(deprecated-data-xml-node), ]  Use <odoo> instead of <odoo><data> or use <odoo noupdate="1">instead of <odoo><data noupdate="1">
pos_multi_session/views/pos_multi_session_views.xml:8: [W7939(deprecated-data-xml-node), ]  Use <odoo> instead of <odoo><data> or use <odoo noupdate="1">instead of <odoo><data noupdate="1">

```

File: pos_multi_session/data/pos_multi_session_data.xml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 Dinar Gabbasov <https://it-projects.info/team/GabbasovDinar>
+     License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html). -->
+
+<odoo>
+
+    <record id="multi_session" model="pos.multi_session">
+        <field name="name">multi session</field>
```
Comment (yelizariev) [inline pos_multi_session/data/pos_multi_session_data.xml]: Rename. Add word *Default*

File: pos_multi_session/demo/demo.xml [hunk 4]
```diff
@@ -6,28 +6,25 @@
...
-            <field name="name">POS1</field>
-            <field name="multi_session_id" ref="demo_multi_session"/>
-        </record>
-        <record id="demo_multi_pos_config_two" model="pos.config">
-            <field name="name">POS2</field>
-            <field name="multi_session_id" ref="demo_multi_session"/>
-        </record>
-        <record id="demo_multi_pos_config_three" model="pos.config">
-            <field name="name">POS3</field>
-            <field name="multi_session_id" ref="demo_multi_session2"/>
```
Comment (yelizariev) [inline pos_multi_session/demo/demo.xml]: Why did you remove demo_multi_session2?

**Review by yelizariev — CHANGES_REQUESTED** [review]
Default Multi-Session must be applied automatically on POS creation

File: pos_multi_session/doc/changelog.rst [hunk 5]
```diff
@@ -1,3 +1,7 @@
+`4.1.0`
+-------
+- **NEW:** A setting in POS config Enable or Disable Synchronization for the POS
```
Comment (yelizariev) [inline pos_multi_session/doc/changelog.rst]: It's not in POS config. Also, add information about required field and explain why we did it

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_multi_session/models/pos_multi_session_models.py [hunk 6]
```diff
@@ -56,6 +56,15 @@ class PosMultiSession(models.Model):
                                  "It's used to prevent synchronization of old orders")
     fiscal_position_ids = fields.Many2many('account.fiscal.position', string='Fiscal Positions', ondelete="restrict")
 
+    @api.multi
+    def action_default_multi_session(self):
+        self.ensure_one()
+        configs = self.env['pos.config'].search([('multi_session_id', '=', False)])
+        for pos in configs:
```
Comment (yelizariev) [inline pos_multi_session/models/pos_multi_session_models.py]: Use just ``configs.write``

File: pos_multi_session/models/pos_multi_session_models.py [hunk 7]
```diff
@@ -56,6 +56,15 @@ class PosMultiSession(models.Model):
                                  "It's used to prevent synchronization of old orders")
     fiscal_position_ids = fields.Many2many('account.fiscal.position', string='Fiscal Positions', ondelete="restrict")
 
+    @api.multi
+    def action_default_multi_session(self):
```
Comment (yelizariev) [inline pos_multi_session/models/pos_multi_session_models.py]: The name doesn't describe what the method does

File: pos_multi_session/models/pos_multi_session_models.py [hunk 8]
```diff
@@ -21,7 +21,7 @@ class PosConfig(models.Model):
                                             'Deselect checkbox of "Active" if this POS should not use syncing.'
                                             'Before updating it you need to close active session',
                                        default=lambda self: self.env.ref('pos_multi_session.default_multi_session', raise_if_not_found=False))
-    multi_session_active = fields.Boolean(string="Active", help="Select the checkbox to enable synchronization for this POS", default=True)
+    multi_session_active = fields.Boolean(string="Active", help="Select the checkbox to enable synchronization for this POS", default=False)
```
Comment (yelizariev) [inline pos_multi_session/models/pos_multi_session_models.py]: Not sure about this. Probably default True is better

File: pos_multi_session/doc/changelog.rst [hunk 9]
```diff
@@ -1,6 +1,6 @@
 `4.1.0`
 -------
-- **NEW:** A setting in POS config Enable or Disable Synchronization for the POS
+- **NEW:** Field "Active" to enable/disable synchronization for POSes. The multi-session parameter is required now to provide compatibility with "POS Multi Session Menu" module
```
Comment (yelizariev) [inline pos_multi_session/doc/changelog.rst]: ... and other modules that put common settings in POS Multi Session

File: pos_multi_session/demo/demo.xml [hunk 10]
```diff
@@ -13,22 +13,24 @@
...
     <record id="demo_multi_pos_config_two" model="pos.config">
         <field name="name">POS2</field>
         <field name="multi_session_id" ref="demo_multi_session"/>
+        <field name="multi_session_active">True</field>
     </record>
 
     <record id="demo_multi_pos_config_three" model="pos.config">
         <field name="name">POS3</field>
         <field name="multi_session_id" ref="default_multi_session"/>
-        <field name="multi_session_active">False</field>
```
Comment (yelizariev) [inline pos_multi_session/demo/demo.xml]: Keep the value even if it's the default one

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_multi_session/demo/demo.xml [hunk 11]
```diff
@@ -9,23 +9,21 @@
 
     <record id="demo_multi_session" model="pos.multi_session">
         <field name="name">Multi Session Demo</field>
+        <field name="active">True</field>
```
Comment (yelizariev) [inline pos_multi_session/demo/demo.xml]: For ``active``  fields odoo hides non-active fields form search results by default

File: pos_multi_session/models/pos_multi_session_models.py [hunk 12]
```diff
@@ -57,13 +57,24 @@ class PosMultiSession(models.Model):
     fiscal_position_ids = fields.Many2many('account.fiscal.position', string='Fiscal Positions', ondelete="restrict")
 
     @api.multi
-    def action_default_multi_session(self):
+    def action_set_default_multi_session(self):
+        """
+            during installation of the module set default multi-session
+            for all POSes for which not specified multi_session_id
```
Comment (yelizariev) [inline pos_multi_session/models/pos_multi_session_models.py]: for all POSes for which multi_session_id is not specified

File: pos_multi_session/models/pos_multi_session_models.py [hunk 13]
```diff
@@ -57,13 +57,24 @@ class PosMultiSession(models.Model):
...
-                'multi_session_id': self.id,
-            })
+        configs.write({
+            'multi_session_id': self.id
+        })
+
+    @api.multi
+    def name_get(self):
+        """ Override name_get method to return generated name."""
+        name = self.name or self._name
```
Comment (yelizariev) [inline pos_multi_session/models/pos_multi_session_models.py]: call super to get base name

**Review by yelizariev — CHANGES_REQUESTED** [review]

**Review by yelizariev — APPROVED** [review]

**Review by ilmir-k — APPROVED** [review]

**Pull request merged**